pkg/instancetype/upgrade.go

```diff
@@ -0,0 +1,144 @@
+//nolint:lll
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2023 Red Hat, Inc.
+ *
+ */
+package instancetype
+
+import (
+	"context"
+
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	virtv1 "kubevirt.io/api/core/v1"
+	instancetypeapi "kubevirt.io/api/instancetype"
+	"kubevirt.io/client-go/log"
+
+	"kubevirt.io/kubevirt/pkg/apimachinery/patch"
+)
+
+type Upgrader interface {
+	Upgrade(vm *virtv1.VirtualMachine) error
+}
+
+func (m *InstancetypeMethods) Upgrade(vm *virtv1.VirtualMachine) error {
+	// Skip if the VirtualMachine doesn't reference an instance type or preference
+	if vm.Spec.Instancetype == nil && vm.Spec.Preference == nil {
+		return nil
+	}
+
+	var (
+		err                        error
+		originalInstancetypeCRName string
+		originalPreferenceCRName   string
+		newInstancetypeCR          *appsv1.ControllerRevision
+		newPreferenceCR            *appsv1.ControllerRevision
+		patchPayload               []byte
+	)
+
+	if vm.Spec.Instancetype != nil && vm.Spec.Instancetype.RevisionName != "" {
+		originalInstancetypeCRName = vm.Spec.Instancetype.RevisionName
+		newInstancetypeCR, patchPayload, err = m.upgradeControllerRevision(vm, vm.Spec.Instancetype.RevisionName, "/spec/instancetype/revisionName")
+		if err != nil {
+			return err
+		}
+	}
+
+	if vm.Spec.Preference != nil && vm.Spec.Preference.RevisionName != "" {
+		originalPreferenceCRName = vm.Spec.Preference.RevisionName
+		var preferencePatchPayload []byte
+		newPreferenceCR, preferencePatchPayload, err = m.upgradeControllerRevision(vm, vm.Spec.Preference.RevisionName, "/spec/preference/revisionName")
+		if err != nil {
+			return err
+		}
+		patchPayload = append(patchPayload, preferencePatchPayload...)
```

Comment:
I believe so but I'll add test coverage.

Reply:
I think merging bytes will not work, because the function that generates payload, `GeneratePatchPayload()`, marshals a slice to json. It's result looks like `[{} , ..., {}]`. Directly appending these json arrays is not a valid json:
`[{} , ..., {}][{} , ..., {}]`